File: packages/go/pkg/bot/chats.go
package bot

import (
	"context"

	"github.com/Nam088/telebot/packages/go/pkg/types"
)

// GetChat gets up to date information about the chat.
//
// Telegram API: https://core.telegram.org/bots/api#getchat
func (b *Bot) GetChat(ctx context.Context, chatID any) (*types.Chat, error) {
	payload := map[string]any{"chat_id": chatID}
	var chat types.Chat
	if err := b.Request(ctx, "getChat", payload, &chat); err != nil {
		return nil, err
	}
	return &chat, nil
}

// GetChatAdministrators gets a list of administrators in a chat.
//
// Telegram API: https://core.telegram.org/bots/api#getchatadministrators
func (b *Bot) GetChatAdministrators(ctx context.Context, chatID any) ([]types.ChatMember, error) {
	payload := map[string]any{"chat_id": chatID}
	var admins []types.ChatMember
	if err := b.Request(ctx, "getChatAdministrators", payload, &admins); err != nil {
		return nil, err
	}
	return admins, nil
}

// GetChatMemberCount gets the number of members in a chat.
//
// Telegram API: https://core.telegram.org/bots/api#getchatmembercount
func (b *Bot) GetChatMemberCount(ctx context.Context, chatID any) (int, error) {
	payload := map[string]any{"chat_id": chatID}
	var count int
	if err := b.Request(ctx, "getChatMemberCount", payload, &count); err != nil {
		return 0, err
	}
	return count, nil
}

// LeaveChat leaves a group, supergroup or channel.
//
// Telegram API: https://core.telegram.org/bots/api#leavechat
func (b *Bot) LeaveChat(ctx context.Context, chatID any) (bool, error) {
	payload := map[string]any{"chat_id": chatID}
	var ok bool
	if err := b.Request(ctx, "leaveChat", payload, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// BanChatMember bans a user in a group, supergroup or channel.
//
// Telegram API: https://core.telegram.org/bots/api#banchatmember
func (b *Bot) BanChatMember(ctx context.Context, chatID any, userID int64, untilDate int64, revokeMessages bool) (bool, error) {
	payload := struct {
		ChatID         any   `json:"chat_id"`
		UserID         int64 `json:"user_id"`
		UntilDate      int64 `json:"until_date,omitempty"`
		RevokeMessages bool  `json:"revoke_messages,omitempty"`
	}{ChatID: chatID, UserID: userID, UntilDate: untilDate, RevokeMessages: revokeMessages}
	var ok bool
	if err := b.Request(ctx, "banChatMember", payload, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// UnbanChatMember unbans a previously banned user in a supergroup or channel.
//
// Telegram API: https://core.telegram.org/bots/api#unbanchatmember
func (b *Bot) UnbanChatMember(ctx context.Context, chatID any, userID int64, onlyIfBanned bool) (bool, error) {
	payload := struct {
		ChatID       any   `json:"chat_id"`
		UserID       int64 `json:"user_id"`
		OnlyIfBanned bool  `json:"only_if_banned,omitempty"`
	}{ChatID: chatID, UserID: userID, OnlyIfBanned: onlyIfBanned}
	var ok bool
	if err := b.Request(ctx, "unbanChatMember", payload, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

File: packages/python/src/telebot_py/bot/chats.py
"""Chat information and moderation Bot API methods (parity with packages/go/pkg/bot/chats.go)."""

from __future__ import annotations

from telebot_py.bot.base import (
    Requester,
    clean_payload,
    parse_count,
    parse_flag,
    parse_list_result,
    parse_result,
)
from telebot_py.types.chat import Chat
from telebot_py.types.chat_members import ChatMember


class ChatsMixin(Requester):
    """Bot methods for chat info, membership counts, and member moderation."""

    async def get_chat(self, chat_id: int | str) -> Chat:
        """Get up-to-date information about a chat.

        Example:
            >>> chat = await bot.get_chat(-1001234567890)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup or channel.

        Returns:
            The Chat object with current chat information.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        return parse_result(Chat, await self.request("getChat", {"chat_id": chat_id}))

    async def get_chat_administrators(self, chat_id: int | str) -> list[ChatMember]:
        """Get a list of administrators in a chat.

        Example:
            >>> admins = await bot.get_chat_administrators(-1001234567890)

        Args:
            chat_id: Unique identifier for the target chat or channel username.

        Returns:
            Every administrator except other bots; empty list if the chat has
            no explicit administrators.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        result = await self.request("getChatAdministrators", {"chat_id": chat_id})
        return parse_list_result(ChatMember, result)

    async def get_chat_member_count(self, chat_id: int | str) -> int:
        """Get the number of members in a chat.

        Example:
            >>> count = await bot.get_chat_member_count(-1001234567890)

        Args:
            chat_id: Unique identifier for the target chat or channel username.

        Returns:
            The member count of the chat.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        return parse_count(await self.request("getChatMemberCount", {"chat_id": chat_id}))

    async def leave_chat(self, chat_id: int | str) -> bool:
        """Leave a group, supergroup, or channel.

        Example:
            >>> await bot.leave_chat(-1001234567890)

        Args:
            chat_id: Unique identifier for the target chat or channel username.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        return parse_flag(await self.request("leaveChat", {"chat_id": chat_id}))

    async def ban_chat_member(
        self,
        chat_id: int | str,
        user_id: int,
        *,
        until_date: int | None = None,
        revoke_messages: bool | None = None,
    ) -> bool:
        """Ban a user in a group, supergroup, or channel.

        Example:
            >>> await bot.ban_chat_member(-1001234567890, 777000)

        Args:
            chat_id: Unique identifier for the target chat or channel username.
            user_id: Unique identifier of the target user.
            until_date: Unix time when the ban ends; ignored if less than 30
                seconds or more than 366 days from now (banned forever).
            revoke_messages: Also delete all messages from the chat for the
                user being banned. Always True for channels.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            user_id=user_id,
            until_date=until_date,
            revoke_messages=revoke_messages,
        )
        return parse_flag(await self.request("banChatMember", payload))

    async def unban_chat_member(
        self,
        chat_id: int | str,
        user_id: int,
        *,
        only_if_banned: bool | None = None,
    ) -> bool:
        """Unban a previously banned user in a supergroup or channel.

        Example:
            >>> await bot.unban_chat_member(-1001234567890, 777000)

        Args:
            chat_id: Unique identifier for the target chat or channel username.
            user_id: Unique identifier of the target user.
            only_if_banned: Do nothing if the user is not banned.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            user_id=user_id,
            only_if_banned=only_if_banned,
        )
        return parse_flag(await self.request("unbanChatMember", payload))
